server/aesdsocket.c

```diff
@@ -14,25 +14,58 @@
 #include <sys/wait.h>
 #include <stdbool.h>
 #include <errno.h>
+#include <pthread.h>
+#include <sys/queue.h>
+#include <time.h>
 
 #define PORT 9000
 #define BACKLOG 10
 #define FILE_PATH "/var/tmp/aesdsocketdata"
 #define BUFFER_SIZE 1024
 
+// SLIST.
+typedef struct slist_data_s slist_data_t;
+struct slist_data_s {
+    pthread_t threadID;  
+    int socket_fd;
+    bool status;
+    int new_fd;
+    struct sockaddr_in *socket_address;
+    SLIST_ENTRY(slist_data_s) entries;
+};
+SLIST_HEAD(slisthead, slist_data_s) head = SLIST_HEAD_INITIALIZER(head);
+
+
 int sockfd;
 volatile sig_atomic_t stop = 0;  // Changed to `sig_atomic_t` for signal safety
+pthread_t timestamp_thread;
+
+#define MAX_THREADS 100
+pthread_t threads[MAX_THREADS];  // Store active thread IDs
+int thread_count = 0;
+pthread_mutex_t aesdsocket_mutex = PTHREAD_MUTEX_INITIALIZER;
+
+struct thread_data
+{
+    int new_fd;
+    struct sockaddr_in *socket_address;
+};
+
 
 void handle_signal(int sig)
 {
     syslog(LOG_INFO, "Caught signal %d, exiting", sig);
+    stop = 1;
     remove(FILE_PATH);
 
     if (sockfd != -1) 
     {
         close(sockfd);
     }
 
+    
+    pthread_join(timestamp_thread, NULL);
+    pthread_mutex_destroy(&aesdsocket_mutex);  // Destroy mutex
     closelog();
     exit(0);
```

Comment:
You won't have cleanup for the last client connection when you exit from your signal handler.  This is what the valgrind warnings are pointing out.